### PR TITLE
Add Custom Java Monitors

### PR DESCRIPTION
**Pull request by roberttoyonaga**
These changes copy the JDK code for ReentrantLock and AbstractQueuedSynchronizer from the JDK19 sources and strip them down to the bare minimum needed by native image for implementing Java monitors. These new bare bones classes are used  to implement Java Monitors instead of their jdk counterparts.

This allows us to have access to the ReentrantLock and AbstractQueuedSynchronizer internals to reduce usage of substitutions and implement JFR monitor events. This approach was described here: https://github.com/oracle/graal/pull/4676#pullrequestreview-1028182562

### CODE REVIEW

File: substratevm/src/com.oracle.svm.core/src/com/oracle/svm/core/monitor/GraalReentrantLock.java [hunk 1]
```diff
@@ -0,0 +1,200 @@
...
+ * questions.
+ */
+
+package com.oracle.svm.core.monitor;
+
+import java.util.concurrent.locks.Condition;
+import com.oracle.svm.core.annotate.Uninterruptible;
+
+/*
+GraalReentrantLock is derived from the ReentrantLock class in the jdk 19 sources. Only the methods required for
```
Comment (christianhaeubl) [inline substratevm/src/com.oracle.svm.core/src/com/oracle/svm/core/monitor/GraalReentrantLock.java:33]: Can be JavaDoc.
Besides that, one of the style checks should complain that there are too many characters per line in this comment.

File: substratevm/src/com.oracle.svm.core/src/com/oracle/svm/core/monitor/GraalAbstractQueuedSynchronizer.java [hunk 2]
```diff
@@ -0,0 +1,672 @@
...
+import java.util.concurrent.locks.LockSupport;
+
+import com.oracle.svm.core.annotate.Alias;
+import com.oracle.svm.core.annotate.TargetClass;
+import com.oracle.svm.core.annotate.TargetElement;
+import com.oracle.svm.core.annotate.Uninterruptible;
+import com.oracle.svm.core.SubstrateUtil;
+
+/*
+GraalAbstractQueuedSynchronizer is derived from the AbstractQueuedSynchronizer class in the jdk 19 sources. Only the methods required for
```
Comment (christianhaeubl) [inline substratevm/src/com.oracle.svm.core/src/com/oracle/svm/core/monitor/GraalAbstractQueuedSynchronizer.java:46]: Please also mention the JDK tag + the git revision that those sources are based on.

File: substratevm/src/com.oracle.svm.core/src/com/oracle/svm/core/monitor/MultiThreadedMonitorSupport.java [hunk 3]
```diff
@@ -488,47 +486,45 @@ protected static JavaMonitor newMonitorLock() {
...
         return false;
     }
 
-    public ReentrantLock getMonitorForTesting(Object obj) {
+    public JavaMonitor getMonitorForTesting(Object obj) {
         return getOrCreateMonitor(obj, true);
     }
 
-    protected ConditionObject getOrCreateCondition(JavaMonitor monitorLock, boolean createIfNotExisting) {
+    protected GraalAbstractQueuedSynchronizer.GraalConditionObject getOrCreateCondition(JavaMonitor monitorLock, boolean createIfNotExisting) {
```
Comment (christianhaeubl) [inline substratevm/src/com.oracle.svm.core/src/com/oracle/svm/core/monitor/MultiThreadedMonitorSupport.java:504]: Feel free to use static imports to make this a bit more readable.

**Review by christianhaeubl — COMMENTED** [review]
Thanks for working on that. Your PR looks already pretty good. I think that there are a few things can be simplified a bit further.

File: substratevm/src/com.oracle.svm.core/src/com/oracle/svm/core/monitor/GraalAbstractQueuedSynchronizer.java [hunk 4]
```diff
@@ -0,0 +1,672 @@
...
+substrateVM monitor support have been kept. Additional functionality specific to substrateVM has been added.
+ */
+public class GraalAbstractQueuedSynchronizer extends AbstractOwnableSynchronizer {
```
[the remark this answers is not in the document]
Comment (roberttoyonaga) [reply substratevm/src/com.oracle.svm.core/src/com/oracle/svm/core/monitor/GraalAbstractQueuedSynchronizer.java]: Ok I removed the inheritance

File: substratevm/src/com.oracle.svm.core/src/com/oracle/svm/core/monitor/GraalAbstractQueuedSynchronizer.java [hunk 5]
```diff
@@ -0,0 +1,672 @@
...
+    }
+
+    public final void acquire(int arg) {
+        if (!tryAcquire(arg)) {
+            acquire(null, arg, false, false, false, 0L);
+        }
+
+    }
+
+    final int acquire(GraalAbstractQueuedSynchronizer.Node node, int arg, boolean shared, boolean interruptible,
```
[the remark this answers is not in the document]
Comment (roberttoyonaga) [reply substratevm/src/com.oracle.svm.core/src/com/oracle/svm/core/monitor/GraalAbstractQueuedSynchronizer.java]: Done!

File: substratevm/src/com.oracle.svm.core/src/com/oracle/svm/core/monitor/GraalReentrantLock.java [hunk 6]
```diff
@@ -0,0 +1,200 @@
...
+substrateVM monitor support have been kept. Additional functionality specific to substrateVM has been added.
+ */
+public class GraalReentrantLock {
```
[the remark this answers is not in the document]
Comment (roberttoyonaga) [reply substratevm/src/com.oracle.svm.core/src/com/oracle/svm/core/monitor/GraalReentrantLock.java]: Renamed using "JavaMonitor" prefix and merged

File: substratevm/src/com.oracle.svm.core/src/com/oracle/svm/core/monitor/JavaMonitor.java [hunk 7]
```diff
@@ -30,20 +30,17 @@
...
 import org.graalvm.nativeimage.CurrentIsolate;
 
-import com.oracle.svm.core.SubstrateUtil;
 import com.oracle.svm.core.jfr.JfrTicks;
 import com.oracle.svm.core.jfr.SubstrateJVM;
 import com.oracle.svm.core.jfr.events.JavaMonitorEnterEvent;
 
-public class JavaMonitor extends ReentrantLock {
+public class JavaMonitor extends GraalReentrantLock {
     private static final long serialVersionUID = 3921577070627519721L;
```
[the remark this answers is not in the document]
Comment (roberttoyonaga) [reply substratevm/src/com.oracle.svm.core/src/com/oracle/svm/core/monitor/JavaMonitor.java]: Removed!

File: substratevm/src/com.oracle.svm.core/src/com/oracle/svm/core/monitor/GraalReentrantLock.java [hunk 8]
```diff
@@ -0,0 +1,200 @@
...
+
+    public GraalAbstractQueuedSynchronizer.GraalConditionObject getCondition() {
+        return sync.getGraalConditionObject();
+    }
+
+    public void setCondition(GraalAbstractQueuedSynchronizer.GraalConditionObject newCondition) {
+        sync.graalConditionObject = newCondition;
+    }
+
+    public static final class GraalNonfairSync extends GraalReentrantLock.Sync {
```
[the remark this answers is not in the document]
Comment (roberttoyonaga) [reply substratevm/src/com.oracle.svm.core/src/com/oracle/svm/core/monitor/GraalReentrantLock.java]: merged classes

File: substratevm/src/com.oracle.svm.core/src/com/oracle/svm/core/monitor/GraalReentrantLock.java (hunk 1 above)
Comment (roberttoyonaga) [reply substratevm/src/com.oracle.svm.core/src/com/oracle/svm/core/monitor/GraalReentrantLock.java]: Ran `mx checkstyle` and also made a JavaDoc

File: substratevm/src/com.oracle.svm.core/src/com/oracle/svm/core/monitor/MultiThreadedMonitorSupport.java (hunk 3 above)
Comment (roberttoyonaga) [reply substratevm/src/com.oracle.svm.core/src/com/oracle/svm/core/monitor/MultiThreadedMonitorSupport.java]: done!

**Comment by christianhaeubl** [comment]
Thanks, I will run some additional testing and will let you know if I encounter any issues.

**Comment by christianhaeubl** [comment]
I did a few cleanups and bugfixes on top of your PR and opened https://github.com/oracle/graal/pull/4731. I also changed the order of all methods and classes so that they match the JDK 19 sources more closely (makes it easier to compare the implementations). I am closing this PR as the other PR includes all your commits.

**Pull request closed without merging**